### PR TITLE
Benja 704 photo upload improvement

### LINKED ISSUE #704: Photo upload form improvements
* "Comment cette photo est-elle en relation avec le quartier des Hauts-Pavés ?"
  --> let's make it clear that this is flexible by giving some examples: j'y habite, j'y travaille, c'est compliqué, autre

* Let's make the place a text field.  Some places are descriptions: the shed behind my house where my aunts walk in strange circles.

* Let's change the recommended resolution to "2000 pixels on the longest side" without changing the hard limit we enforce in software.

* Let's change the order:
  - category
  - photo
  - lieu
  - commentaire du photographe
  - hauts-pavés
  - appareil
  - notes techniques

* Before validation, we need a checkbox that says "I've read and abided by `[T&C](operation-pieton-participer)` below" which we should then copy from operation-pieton-participer in a less chatty way.  Feel free to improve the French, suggest rewordings to be clearer or more precise:

```
Je certifie que
* Je comprends que la compétition n'aura pas de récompense pécuniaire, ni aux participants ni aux gagnants.
* Ma photo représente un élément de l'expérience piétonne.
* Je ne suis pas entré en infraction pour prendre la photo.
* La diffusion de l'image ne crée pas une infraction.
* La photo respecte le droit d'image et le droit de panorama d'autrui.
* Je reconnais que je cède gratuitement, pendant trois ans, un droit de diffusion non exclusif aux Mobilitains pour l'exploitation de la compétition ainsi que pour la diffusion des images retenues, et dans le contexte formel d'une exposition et dans le contexte informel du site web des Mobilitains qui afficherait les lauréats et mentions.  Mon image pourrait également être diffusé sur réseaux sociaux dans ces mêmes buts.
* J'atteste que l'essentiel de la photo soumise est la mienne, que toute manipulation éventuelle est technique, pas l'effet principal.
```

### PR DESCRIPTION
**Pull request by Shriukan33**
This PR implements the changes proposed in issue #704 : 
* Add the placeholder in the relationship with Haut-pavés
* Make the location a text field
* Change the recommended width of the photo
* Reorder the fields
* Add ToS checkbox

Closes #704

### CODE REVIEW

**Comment by JeffAbrahamson** [comment]
Issue #707 seems to be just bug fixes on issues #696 (which remains open).

**Comment by Shriukan33** [comment]
Indeed, should we close both then ?

**Comment by JeffAbrahamson** [comment]
a51f00e - I think you are modifying `UploadEntry`, not `CreateView`.  It might be clearer to say "Use a custom for class in UploadEntry" rather than focusing on making CreateView do something.

**Comment by JeffAbrahamson** [comment]
3af71f11 - I don't think you need to define what a placeholder is, that's common web terminology, even if (it appears ;-) you've just learned about it.

It would make more sense to precede the placeholder with something like "Exemple :" to make clear that you are providing an example and that the field is not already filled in.

**Comment by Shriukan33** [comment]
> [3af71f1](https://github.com/transport-nantes/tn_web/commit/3af71f117b5b5bdeb57408d10221506f55c4a34d) - I don't think you need to define what a placeholder is, that's common web terminology, even if (it appears ;-) you've just learned about it.
> 
> It would make more sense to precede the placeholder with something like "Exemple :" to make clear that you are providing an example and that the field is not already filled in.

Haha, I definitely not learned it today, but I preferred giving too much explanations than not enough, I know you can be picky sometimes :D 
Will add "Examples : " in the placeholder :)

**Comment by JeffAbrahamson** [comment]
cf2159a3 - you have an absolute URL!!

**Comment by Shriukan33** [comment]
> [cf2159a](https://github.com/transport-nantes/tn_web/commit/cf2159a3ece969d0227997e533183b6bb67ea0c7) - you have an absolute URL!!

We already published that one, figured it wasn't at risk to change so used the absolute URL. 
Would you rather see a reverse here ?

**Comment by JeffAbrahamson** [comment]
> > [cf2159a](https://github.com/transport-nantes/tn_web/commit/cf2159a3ece969d0227997e533183b6bb67ea0c7) - you have an absolute URL!!
> 
> We already published that one, figured it wasn't at risk to change so used the absolute URL. Would you rather see a reverse here ?

You should never use absolute URLs in your django app to refer to your own site.  You should always use URL reverse resolution.

For example, what you've done here means that you point to production, even when testing in beta or dev.  If that page should have some behaviour or different content, or if one is developing offline, or if some day we do something every so slightly more generic, it will break.

**Comment by JeffAbrahamson** [comment]
There's a stray bullet at the end.

image

**Comment by Shriukan33** [comment]
> There's a stray bullet at the end.
> 
> image

Not a stray bullet, it's the * indicating mandatory fields

**Comment by JeffAbrahamson** [comment]
"Commentaires du photographe" - put your self in the user's shoes.  Are these "commentaires du photographe" or are they "vos commentaires sur la photo", for example?

**Comment by JeffAbrahamson** [comment]
The image size gate is set to 1000x1000, but we've said "on the longest side".  So an image that's, say, 800 x 1200, would fail.  Can we be more lenient without accepting really tiny stuff?

We can always not select stuff later for being too small...

**Comment by JeffAbrahamson** [comment]
LOL.  This is the idea, but maybe we could demonstrate that we are more literate than this!  First, we could make the text more friendly.  Something about participating in a community project and expressing that you think good mobility is a part of all of our lives.

And then pitch another photo.  I can have a think about this, it needn't hold us up.

image

**Comment by Shriukan33** [comment]
> > > [cf2159a](https://github.com/transport-nantes/tn_web/commit/cf2159a3ece969d0227997e533183b6bb67ea0c7) - you have an absolute URL!!
> > 
> > 
> > We already published that one, figured it wasn't at risk to change so used the absolute URL. Would you rather see a reverse here ?
> 
> You should never use absolute URLs in your django app to refer to your own site. You should always use URL reverse resolution.
> 
> For example, what you've done here means that you point to production, even when testing in beta or dev. If that page should have some behaviour or different content, or if one is developing offline, or if some day we do something every so slightly more generic, it will break.

Turns out doing a reverse_lazy in the form wont work because it needs to be immediately evaluated to instantiate the form field, and Django raises this error : 
```
 Traceback (most recent call last):
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/utils/autoreload.py", line 64, in wrapper
    fn(*args, **kwargs)
  File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/core/management/commands/runserver.py", line 118, in inner_run
    self.check(display_num_errors=True)
  File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/core/management/base.py", line 419, in check
    all_issues = checks.run_checks(
  File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/core/checks/registry.py", line 76, in run_checks
    new_errors = check(app_configs=app_configs, databases=databases)
  File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/core/checks/urls.py", line 13, in check_url_config
    return check_resolver(resolver)
  File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/core/checks/urls.py", line 23, in check_resolver
    return check_method()
  File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/urls/resolvers.py", line 416, in check
    for pattern in self.url_patterns:
  File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/utils/functional.py", line 48, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/urls/resolvers.py", line 602, in url_patterns
    patterns = getattr(self.urlconf_module, "urlpatterns", self.urlconf_module)
  File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/utils/functional.py", line 48, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/urls/resolvers.py", line 595, in urlconf_module
    return import_module(self.urlconf_name)
  File "/usr/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 848, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/benja/Documents/GitHub/tn_web/transport_nantes/transport_nantes/urls.py", line 52, in <module>
    path("photo/", include('photo.urls')),
  File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/urls/conf.py", line 34, in include
    urlconf_module = import_module(urlconf_module)
  File "/usr/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 848, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/benja/Documents/GitHub/tn_web/transport_nantes/photo/urls.py", line 2, in <module>
    from . import views
  File "/home/benja/Documents/GitHub/tn_web/transport_nantes/photo/views.py", line 6, in <module>
    from .forms import PhotoEntryForm
  File "/home/benja/Documents/GitHub/tn_web/transport_nantes/photo/forms.py", line 6, in <module>
    class PhotoEntryForm(forms.ModelForm):
  File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/forms/models.py", line 276, in __new__
    raise FieldError(message)
django.core.exceptions.FieldError: Unknown field(s) (terms_and_condition_checkbox) specified for PhotoEntry
```

This is because it tries to evaluate the reverse even though it's lazy because it's inserted in the text.
At the moment  it evaluates this, the urlpattern seems to be empty, or not initialized yet.

**Comment by JeffAbrahamson** [comment]
> > [3af71f1](https://github.com/transport-nantes/tn_web/commit/3af71f117b5b5bdeb57408d10221506f55c4a34d) - I don't think you need to define what a placeholder is, that's common web terminology, even if (it appears ;-) you've just learned about it.
> > It would make more sense to precede the placeholder with something like "Exemple :" to make clear that you are providing an example and that the field is not already filled in.
> 
> Haha, I definitely not learned it today, but I preferred giving too much explanations than not enough, I know you can be picky sometimes :D Will add "Examples : " in the placeholder :)

Used a radiobutton.  Radiobuttons permit a choice of precisely one multiple options.  Followed by a text field.  Text fields allow text with embedded new lines.  ;-)

**Comment by JeffAbrahamson** [comment]
> There's a stray bullet at the end.
> 
> image

Ah, so it is!  It doesn't read that way, sadly...

Perhaps have the checkbox line, then push the bullet list into `photoentry_form.html`?

**Comment by JeffAbrahamson** [comment]
> > > > [cf2159a](https://github.com/transport-nantes/tn_web/commit/cf2159a3ece969d0227997e533183b6bb67ea0c7) - you have an absolute URL!!
> > > 
> > > 
> > > We already published that one, figured it wasn't at risk to change so used the absolute URL. Would you rather see a reverse here ?
> > 
> > 
> > You should never use absolute URLs in your django app to refer to your own site. You should always use URL reverse resolution.
> > For example, what you've done here means that you point to production, even when testing in beta or dev. If that page should have some behaviour or different content, or if one is developing offline, or if some day we do something every so slightly more generic, it will break.
> 
> Turns out doing a reverse_lazy in the form wont work because it needs to be immediately evaluated to instantiate the form field, and Django raises this error :
> 
> ```
>  Traceback (most recent call last):
>   File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
>     self.run()
>   File "/usr/lib/python3.8/threading.py", line 870, in run
>     self._target(*self._args, **self._kwargs)
>   File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/utils/autoreload.py", line 64, in wrapper
>     fn(*args, **kwargs)
>   File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/core/management/commands/runserver.py", line 118, in inner_run
>     self.check(display_num_errors=True)
>   File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/core/management/base.py", line 419, in check
>     all_issues = checks.run_checks(
>   File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/core/checks/registry.py", line 76, in run_checks
>     new_errors = check(app_configs=app_configs, databases=databases)
>   File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/core/checks/urls.py", line 13, in check_url_config
>     return check_resolver(resolver)
>   File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/core/checks/urls.py", line 23, in check_resolver
>     return check_method()
>   File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/urls/resolvers.py", line 416, in check
>     for pattern in self.url_patterns:
>   File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/utils/functional.py", line 48, in __get__
>     res = instance.__dict__[self.name] = self.func(instance)
>   File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/urls/resolvers.py", line 602, in url_patterns
>     patterns = getattr(self.urlconf_module, "urlpatterns", self.urlconf_module)
>   File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/utils/functional.py", line 48, in __get__
>     res = instance.__dict__[self.name] = self.func(instance)
>   File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/urls/resolvers.py", line 595, in urlconf_module
>     return import_module(self.urlconf_name)
>   File "/usr/lib/python3.8/importlib/__init__.py", line 127, in import_module
>     return _bootstrap._gcd_import(name[level:], package, level)
>   File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
>   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
>   File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
>   File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
>   File "<frozen importlib._bootstrap_external>", line 848, in exec_module
>   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
>   File "/home/benja/Documents/GitHub/tn_web/transport_nantes/transport_nantes/urls.py", line 52, in <module>
>     path("photo/", include('photo.urls')),
>   File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/urls/conf.py", line 34, in include
>     urlconf_module = import_module(urlconf_module)
>   File "/usr/lib/python3.8/importlib/__init__.py", line 127, in import_module
>     return _bootstrap._gcd_import(name[level:], package, level)
>   File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
>   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
>   File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
>   File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
>   File "<frozen importlib._bootstrap_external>", line 848, in exec_module
>   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
>   File "/home/benja/Documents/GitHub/tn_web/transport_nantes/photo/urls.py", line 2, in <module>
>     from . import views
>   File "/home/benja/Documents/GitHub/tn_web/transport_nantes/photo/views.py", line 6, in <module>
>     from .forms import PhotoEntryForm
>   File "/home/benja/Documents/GitHub/tn_web/transport_nantes/photo/forms.py", line 6, in <module>
>     class PhotoEntryForm(forms.ModelForm):
>   File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/forms/models.py", line 276, in __new__
>     raise FieldError(message)
> django.core.exceptions.FieldError: Unknown field(s) (terms_and_condition_checkbox) specified for PhotoEntry
> ```
> 
> This is because it tries to evaluate the reverse even though it's lazy because it's inserted in the text. At the moment it evaluates this, the urlpattern seems to be empty, or not initialized yet.

Fair enough.

Of course, this is python, you could compute the text complete with calling functions like `reverse()`, then assign that text to the label...

**Comment by Shriukan33** [comment]
> > There's a stray bullet at the end.
> > image
> 
> Ah, so it is! It doesn't read that way, sadly...
> 
> Perhaps have the checkbox line, then push the bullet list into `photoentry_form.html`?

Will do that

> > > [3af71f1](https://github.com/transport-nantes/tn_web/commit/3af71f117b5b5bdeb57408d10221506f55c4a34d) - I don't think you need to define what a placeholder is, that's common web terminology, even if (it appears ;-) you've just learned about it.
> > > It would make more sense to precede the placeholder with something like "Exemple :" to make clear that you are providing an example and that the field is not already filled in.
> > 
> > 
> > Haha, I definitely not learned it today, but I preferred giving too much explanations than not enough, I know you can be picky sometimes :D Will add "Examples : " in the placeholder :)
> 
> Used a radiobutton. Radiobuttons permit a choice of precisely one multiple options. Followed by a text field. Text fields allow text with embedded new lines. ;-)

So I add a new field in the model and display its choices as radio ? What's the point compared to the text we have ?

**Comment by JeffAbrahamson** [comment]
> > Used a radiobutton. Radiobuttons permit a choice of precisely one multiple options. Followed by a text field. Text fields allow text with embedded new lines. ;-)
> 
> So I add a new field in the model and display its choices as radio ? What's the point compared to the text we have ?

No!  I'm teasing you about having commit comments that define command terminology.

**Comment by Shriukan33** [comment]
> > > > > [cf2159a](https://github.com/transport-nantes/tn_web/commit/cf2159a3ece969d0227997e533183b6bb67ea0c7) - you have an absolute URL!!
> > > > 
> > > > 
> > > > We already published that one, figured it wasn't at risk to change so used the absolute URL. Would you rather see a reverse here ?
> > > 
> > > 
> > > You should never use absolute URLs in your django app to refer to your own site. You should always use URL reverse resolution.
> > > For example, what you've done here means that you point to production, even when testing in beta or dev. If that page should have some behaviour or different content, or if one is developing offline, or if some day we do something every so slightly more generic, it will break.
> > 
> > 
> > Turns out doing a reverse_lazy in the form wont work because it needs to be immediately evaluated to instantiate the form field, and Django raises this error :
> > ```
> >  Traceback (most recent call last):
> >   File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
> >     self.run()
> >   File "/usr/lib/python3.8/threading.py", line 870, in run
> >     self._target(*self._args, **self._kwargs)
> >   File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/utils/autoreload.py", line 64, in wrapper
> >     fn(*args, **kwargs)
> >   File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/core/management/commands/runserver.py", line 118, in inner_run
> >     self.check(display_num_errors=True)
> >   File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/core/management/base.py", line 419, in check
> >     all_issues = checks.run_checks(
> >   File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/core/checks/registry.py", line 76, in run_checks
> >     new_errors = check(app_configs=app_configs, databases=databases)
> >   File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/core/checks/urls.py", line 13, in check_url_config
> >     return check_resolver(resolver)
> >   File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/core/checks/urls.py", line 23, in check_resolver
> >     return check_method()
> >   File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/urls/resolvers.py", line 416, in check
> >     for pattern in self.url_patterns:
> >   File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/utils/functional.py", line 48, in __get__
> >     res = instance.__dict__[self.name] = self.func(instance)
> >   File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/urls/resolvers.py", line 602, in url_patterns
> >     patterns = getattr(self.urlconf_module, "urlpatterns", self.urlconf_module)
> >   File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/utils/functional.py", line 48, in __get__
> >     res = instance.__dict__[self.name] = self.func(instance)
> >   File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/urls/resolvers.py", line 595, in urlconf_module
> >     return import_module(self.urlconf_name)
> >   File "/usr/lib/python3.8/importlib/__init__.py", line 127, in import_module
> >     return _bootstrap._gcd_import(name[level:], package, level)
> >   File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
> >   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
> >   File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
> >   File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
> >   File "<frozen importlib._bootstrap_external>", line 848, in exec_module
> >   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
> >   File "/home/benja/Documents/GitHub/tn_web/transport_nantes/transport_nantes/urls.py", line 52, in <module>
> >     path("photo/", include('photo.urls')),
> >   File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/urls/conf.py", line 34, in include
> >     urlconf_module = import_module(urlconf_module)
> >   File "/usr/lib/python3.8/importlib/__init__.py", line 127, in import_module
> >     return _bootstrap._gcd_import(name[level:], package, level)
> >   File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
> >   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
> >   File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
> >   File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
> >   File "<frozen importlib._bootstrap_external>", line 848, in exec_module
> >   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
> >   File "/home/benja/Documents/GitHub/tn_web/transport_nantes/photo/urls.py", line 2, in <module>
> >     from . import views
> >   File "/home/benja/Documents/GitHub/tn_web/transport_nantes/photo/views.py", line 6, in <module>
> >     from .forms import PhotoEntryForm
> >   File "/home/benja/Documents/GitHub/tn_web/transport_nantes/photo/forms.py", line 6, in <module>
> >     class PhotoEntryForm(forms.ModelForm):
> >   File "/home/benja/Documents/GitHub/tn_web/venv/lib/python3.8/site-packages/django/forms/models.py", line 276, in __new__
> >     raise FieldError(message)
> > django.core.exceptions.FieldError: Unknown field(s) (terms_and_condition_checkbox) specified for PhotoEntry
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This is because it tries to evaluate the reverse even though it's lazy because it's inserted in the text. At the moment it evaluates this, the urlpattern seems to be empty, or not initialized yet.
> 
> Fair enough.
> 
> Of course, this is python, you could compute the text complete with calling functions like `reverse()`, then assign that text to the label...

I don't understand your suggestion

**Comment by JeffAbrahamson** [comment]
> > Fair enough.
> > Of course, this is python, you could compute the text complete with calling functions like `reverse()`, then assign that text to the label...
> 
> I don't understand your suggestion

Something like  this:

```
class PhotoEntryForm(forms.ModelForm):

    t_and_c_url = reverse("topicblog.view_item_by_slug",
                          "operation-pieton-participer")
    the_label = f"""
    J'accepte les
        <a href="{ t_and_c_url }">
        conditions générales d'utilisation</a> et certifie que :
    """
    terms_and_condition_checkbox = forms.BooleanField(
        label=the_label, required=True,
    )

    class Meta:
        ...
```

and

```
{% block app_content %}
<div class="container">
    <div class="row">
        <div class="col-lg-12 pb-3">
            {# ####  This should be db-driven!  #### #}
            <h1>Formulaire de participation à l'Opération piéton</h1>
            <form action="{% url 'photo:upload' %}" method="POST" enctype="multipart/form-data">
                {% csrf_token %}
                {{ form|crispy }}

		<div ...>
		    <p>
			<ul>
        <li>Je comprends que la compétition n'aura pas de récompense
        pécuniaire, ni aux participants ni aux gagnants.</li>
        <li>Ma photo représente un élément de l'expérience piétonne.</li>
        <li>Je ne suis pas entré en infraction pour prendre la photo.</li>
        <li>La diffusion de l'image ne crée pas une infraction.</li>
        <li>La photo respecte le droit d'image et le droit de panorama
        d'autrui.</li>
        <li>Je reconnais que je cède gratuitement, pendant trois ans, un droit
        de diffusion non exclusif aux Mobilitains pour l'exploitation de la
        compétition ainsi que pour la diffusion des images retenues, et dans
        le contexte formel d'une exposition et dans le contexte informel du
        site web des Mobilitains qui afficherait les lauréats et mentions.
        Mon image pourrait également être diffusé sur réseaux sociaux dans ces
        mêmes buts.</li>
        <li>J'atteste que l'essentiel de la photo soumise est la mien,
        que toute manipulation éventuelle est technique, pas l'effet
        principal.</li>
        </ul>
		    </p>
		</div>
                <button class="btn donation-button" type="submit">Valider</button>
            </form>
        </div>
    </div>

</div>

{% endblock %}
```

**Comment by JeffAbrahamson** [comment]
LGTM.
